svthalia/concrexit
# Face detection (with lambda)

Supersedes #2603

### Summary
Implements face detection as a separate django app in concrexit, making use of [svthalia/concrexit-facedetection-lambda](https://github.com/svthalia/concrexit-facedetection-lambda) to perform extraction of face encodings.

### How to test
You can test this locally, but that's not trivial:
1. Build a slightly modified lambda image, adding `verify=False` in https://github.com/svthalia/concrexit-facedetection-lambda/blob/6a254af7298d5b37ac73052d1fac0151cb53fdde/app.py#L27.
  This will prevent SSLErrors in the lambda when working with a local concrexit.
2. Replace https://github.com/svthalia/concrexit/blob/76eda9504c2e0188a880c817de2810c95fcfc035/website/facedetection/services.py#L74-L88 with:

```python
        response = requests.post(
            settings.FACEDETECTION_LAMBDA_ARN,
            json=payload,
        )

        if response.status_code != 200:
            raise Exception("Couldn't invoke Lambda function by url.")
```
3. Run the modified docker image with `docker run -p 9000:8080 concrexit-face-detection-lambda`
4. Run concrexit with `FACEDETECTION_LAMBDA_ARN="http://localhost:9000/2015-03-31/functions/function/invocations" BASE_URL="http://host.docker.internal:8000" ./manage.py runserver`

5. Now concrexit should be working with your local lambda emulator. Might take some effort to figure out the details though. I don't expect we'll be touching any of this a lot later, so I ended up not including support for local running in the code. For normal local development just never extracting encodings seems fine to me.

Anyway, the only way to really be sure this works would be to test it with a real lambda. So we can pretty much just try it on staging when the code looks good. 

### TODO:
- [x] Triggering analysis of albums/photos (Made a nightly management command, but I would like a way to also trigger it after an album has been uploaded.)
- [x] Add trigger after album has been uploaded (a custom signal sent from the album admin after saving is done?)
- [x] What to do when there is no lambda configured?... (It's logged but ignored. Same goes for failures, so in the lambda being down shouldn't ever crash concrexit).
- [x] A readme documenting the facedetection app
- [x] Modify media url expiration time to allow longer queuing (lambda calls may be queud for up to 6 hours by default, but right now after 1 hour the urls will break)
